beginning toc and filesystem reconfig (#2790)
getporter/porter: docs/content/introduction/intro-invocation-images.md

---

> title: Invocation Images

Reviewer: ```suggestion
title: Bundle Installers
```
[Carolyn wanted this change awhile back](https://github.com/getporter/porter/issues/1188) and honestly I can't think of a better name

> descriptions: How does Porter build an Invocation Image?

Reviewer: ```suggestion
descriptions: How does Porter build an Bundle Installer?
```

---

> When you build a Cloud Native Application Bundle (CNAB) with Porter, a bundle.json and an invocation image are created for you. How does Porter turn your _porter.yaml_ into an invocation image? This walkthrough will explain how Porter constructs the invocation image, including how mixins and other bundles allow you to compose functionality.

Reviewer: ```suggestion
When you build a Cloud Native Application Bundle (CNAB) with Porter, a bundle.json and a bundle installer is created for you. How does Porter turn your _porter.yaml_ into a bundle installer? This walkthrough will explain how Porter constructs the bundle installer, including how mixins and other bundles allow you to compose functionality.
```

## Starting From Scratch

> When you create a new bundle with Porter, your project is bootstrapped with a sample porter.yaml. This scaffolding provides almost everything you need to generate your CNAB, including the invocation image. Let's use this to explain how the invocation image is built.

Reviewer: ```suggestion
When you create a new bundle with Porter, your project is bootstrapped with a sample porter.yaml. This scaffolding provides almost everything you need to generate your CNAB, including the bundle installer. Let's use this to explain how the bundle installer is built.
```

Copying mixin exec ===>
```

First, Porter copies its runtime plus any mixins into the `.cnab/app` directory of your bundle.

> Porter locates available mixins in the $PORTER_HOME/mixins directory. By default, the Porter home directory is located in ~/.porter. In this example, we are using the exec mixin, so the $PORTER_HOME/mixins/exec directory will be copied into the invocation image. When a mixin is installed, it contains binaries for multiple operating systems. The correct binary will be copied into the bundle's .cnab directory for use in the invocation image.

Reviewer: ```suggestion
Porter locates available mixins in the $PORTER_HOME/mixins directory. By default, the Porter home directory is located in ~/.porter. In this example, we are using the exec mixin, so the $PORTER_HOME/mixins/exec directory will be copied into the bundle installer. When a mixin is installed, it contains binaries for multiple operating systems. The correct binary will be copied into the bundle's .cnab directory for use in the bundle installer.
```

CMD ["/cnab/app/run"]
```

> Porter starts the [Dockerfile](/bundle/custom-dockerfile) by using a base image. You can customize the base image by specifying a Dockerfile template in the porter.yaml. By default, Porter only targets a single os/architecture(linux/amd64) for invocation image. If you want to use other platform, feel free to change the platform flag in the generated Dockerfile template. Next, a set of CA certificates is added. Next, contents of the current directory are copied into the bundle directory (/cnab/app) in the invocation image. This will include any contributions from the mixin executables. Finally, an entry point that conforms to the CNAB specification is added to the image.

Reviewer: ```suggestion
Porter starts the [Dockerfile](/bundle/custom-dockerfile) by using a base image. You can customize the base image by specifying a Dockerfile template in the porter.yaml. By default, Porter only targets a single os/architecture(linux/amd64) for the bundle installer. If you want to use other platform, feel free to change the platform flag in the generated Dockerfile template. Next, a set of CA certificates is added. Then, contents of the current directory are copied into the bundle directory (/cnab/app) in the bundle installer. This will include any contributions from the mixin executables. Finally, an entry point that conforms to the CNAB specification is added to the image.
```

Porter mixins must provide a build command that generates Dockerfile lines to support the runtime execution of the mixin. In the case of the helm mixin, this includes installing Helm and running a `helm init --client-only` to prepare the image. At build time, Porter uses the porter.yaml to determine what mixins are required for the bundle. Then Porter invokes the build sub-command for each specified mixin and appends that output to the base Dockerfile.

> In the end, the result is a single invocation image with the necessary pieces: the porter-runtime, selected mixins and any relevant configuration files, scripts, charts or manifests. That invocation image can then be executed by any tool that supports the CNAB spec, while still taking advantage of the Porter capabilities.

Reviewer: ```suggestion
In the end, the result is a single bundle installer with the necessary pieces: the porter-runtime, selected mixins and any relevant configuration files, scripts, charts or manifests. That bundle installer can then be executed by any tool that supports the CNAB spec, while still taking advantage of the Porter capabilities.
```
